README.md

Greetings! This project offers monitoring and alerting of Chia (and other forks) farms.  Designed to be **bundled** with the powerful [Machinaris](https://www.machinaris.app) plotting and farming tool, this fork of the [original Chiadog](https://github.com/martomi/chiadog) adds the following capabilities:

1. Monitoring of blockchain forks such as Flax, HDDCoin, Chives, etc.  Not just Chia!
2. Enhanced configuration syntax to selectively allow or ignore alerts on each notifier.
3. An API endpoint for other applications to send their alerts to the same Chiadog notification targets.

The following documentation below is from the original Chiadog project.  Please ignore the installation steps for Chiadog as that is taken care of by simply installing [Machinaris](https://www.machinaris.app).  Thanks!

## Supported Notifications

| Subsystem | Notification (example values) | Priority |
| ------------- | ------------- | ------|
| Harvester | Your harvester appears to be offline! No events for the past 400 seconds. | HIGH |
| Harvester | Disconnected HDD? The total plot count decreased from 100 to 40. | HIGH |
| Harvester | Connected HDD? The total plot count increased from 0 to 42. | LOW |
| Harvester | Experiencing networking issues? Harvester did not participate in any challenge for 120 seconds. It's now working again. | NORMAL |
| Harvester | Seeking plots took too long: 21.42 seconds! | NORMAL |
| Full Node | Experiencing networking issues? Skipped 42 signage points! | NORMAL |
| Full Node | Block found!! | LOW |
| Wallet | Cha-ching! Just received 2.0 XCH ☘️ | LOW |
| Daily Stats | Hello farmer! 👋 Here's what happened in the last 24 hours: <br /><br /> Received ☘️: **2.00** XCH️<br /> Proofs 🧾: **176** found!<br />  - **176** partials submitted 📑<br /> - **0** blocks found 📦<br/> Search 🔍: <br /> - average: **0.46**s <br /> - over 5s: 2 occasions <br /> - over 15s: 1 occasions <br/> Plots 🌱: **42**, new: **2** <br /> Eligible plots 🥇: **0.08** average<br /> Skipped SPs ⚠️: 7 (0.01%) <br /> | LOW |

Please refer to [Status Reference](https://github.com/martomi/chiadog/wiki/Status-Reference) page for detailed
explanations of the notifications.

## How it works?

It parses the `debug.log` generated by the chia process and runs various checks to determine the health of your farmer.
Among others, it can detect if your node has lost sync and the farmer is no longer participating in challenges, or if
one of your external HDDs disconnected and your harvester doesn't have access to the full amount of plots you have.

## Access & Security

It only requires read-access to your `debug.log` file and internet connection to send out notifications. It's highly
recommended that you run `chiadog` in a sandboxed environment. Please use the official [docker image](https://github.com/martomi/chiadog/pkgs/container/chiadog/versions).

Furthermore, following best security practices, you
should [keep your wallet separate](https://github.com/Chia-Network/chia-blockchain/wiki/Good-Security-Practices-on-Many-Machines#keep-your-wallet-separate).

## Supported Integrations for Notifications

You may use one (or more) of the following integrations to receive notifications from `chiadog`.

| Integration | Advantages | Cost |
| ------------- | ------------- | ------|
| [Pushover](https://pushover.net/) | High priority notifications that can override your phone's silent mode. | $5 one time purchase after 30 day trial. |
| [Pushcut](https://pushcut.io/) | Alternative to Pushover |
| E-mail | You probably already have an email. No additional apps. | Free |
| Slack | Quick & easy setup. | Free |
| Discord | Quick & easy setup. | Free |
| Telegram | Quick & easy setup. | Free |
| Shell script (beta) | Execute anything in your own script. | Free |
| MQTT | Well-suited for Home Automation. | Free | 
| Grafana | For hardware monitoring. | Free |
| [Ifttt](https://ifttt.com/) | Can be used to send push notifications or to do other API integrations depending on incoming data. | Free |


For detailed guide on how to test and configure, please refer to [INTEGRATIONS.md](INTEGRATIONS.md).

# Getting started

## Pre-requisites

- Linux, MacOS & Windows
- Python 3.7+
- [Git](https://git-scm.com/downloads)
- Enabled `INFO` logs on your chia farmer

The instructions below are specific to Linux and MacOS, for installing `chiadog` on Windows, please refer to this
separate [README](WINDOWS.md) section.

### How to enable INFO logs on chia farmer?

First configure the log level to `INFO`. This ensures that all logs necessary for chiadog to operate are available
under `~/.chia/mainnet/log/debug.log`.

```
chia configure -log-level=INFO
```

Then restart your farmer to apply the changes:

```
chia start --restart farmer
```

Check that logs are coming in:

```
cat ~/.chia/mainnet/log/debug.log
```

## Installation

### Recommended

The new recommended way of using `chiadog` is via the official [docker image](https://github.com/martomi/chiadog/pkgs/container/chiadog/versions).

### Manual Installation

_For updating from previous version, see section below._

1. Clone the repository

```
git clone https://github.com/martomi/chiadog.git
cd chiadog
```

2. Run the install script.

```
./install.sh
```

3. Copy the example config file

```
cp config-example.yaml config.yaml
```

4. Open up `config.yaml` in your editor and configure it to your preferences.

### Updating to the latest release

_Skip this if you followed the above section_.

```
cd chiadog

git fetch
git checkout main
git pull

./install.sh
```

> Important: Automated migration of config is not supported. Please check that your `config.yaml` has all new fields introduced in `config-example.yaml` and add anything missing. If correctly migrated, you shouldn't get any ERROR logs.

## Monitoring a local harvester / farmer

1. Open `config.yaml` and configure `file_log_consumer`:
    - You need to enable the file log consumer to read local chia log files
    - Double-check that the path to your chia logs is correct

2. Start the watchdog

```
./start.sh
```

3. Verify that your plots are detected. Within a few seconds you should see INFO log:

```
Detected new plots. Farming with 42 plots.
```

If you see the above log message, you can be certain that `chiadog` is running correctly. It'll remain silent until the
next scheduled daily notification or until any issues are detected.

You can repeat the above process for every machine where you are running a harvester. Use
the `notification_title_prefix`
in `config.yaml` to give every machine a unique notification prefix so that you can easily distinguish them.

If you don't want to setup `chiadog` on each machine separately, you can also monitor multiple remote harvesters and run
chiadog on a single machine. Please refer to the wiki page
on [Remote Monitoring Multiple Harvesters](https://github.com/martomi/chiadog/wiki/Monitoring-Multiple-Harvesters).

## Troubleshooting

You can enable more verbose logging from `config.yaml` by changing `INFO` to `DEBUG`. You should see logs for every
keep-alive event from the harvester.

# Advanced Usage

## Redundant monitoring for `chiadog`

There are failure-cases in which `chiadog` is helpless. For example, your computer completely freezes or shuts down.
Perhaps your entire home network goes down. `chiadog` won't be able to send you a notification.

There's a way however: in the [config](config-example.yaml) under the section of `keep_alive_monitor`, you can enable
pinging to a remote service that will act as a watchdog of `chiadog`. A second level of redundancy, if you wish!

You may chose your favourite service for that, I've tested it with
[HealthChecks.io](https://healthchecks.io). It's free to signup and create an endpoint that expects to receive pings
every 10 minutes. If it does not, it will notify you. It has integrations with Pushover, Email, Slack, Discord and more.

## Running `chiadog` in the background

```
. ./venv/bin/activate
nohup python3 -u main.py --config config.yaml > output.log 2>&1 &
```

To stop chiadog, you can find the Process ID (PID) via `ps aux | grep main.py` and then softly interrupt the process
with `kill -SIGINT <pid_here>`.

## Running `chiadog` as sandboxed systemd service

Alternatively to the original chiadog docker image, you can setup a [systemd service](scripts/linux/chiadog.service)
which runs chiadog as a limited user and blocks access to key chia locations.

# Contributing

Contributions are always welcome! Please refer to [CONTRIBUTING](CONTRIBUTING.md) documentation.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres
to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.7.5] - 2022-11-05

### Fixed
- Fixed wallet parser for Chia v1.6.1 - thanks [@djerfy](https://github.com/djerfy)!

## [0.7.4] - 2022-11-02

### Fixed
- Bumped `paramiko` dependency to latest version (2.11.0) to fix a deprecation warning

## [0.7.3] - 2022-08-26

### Fixed
- Fixed wallet parser for Chia v1.5.1+ - thanks [@djerfy](https://github.com/djerfy)!

## [0.7.2] - 2022-07-04

### Fixed
- Fixed wallet parser for Chia v1.4.0+ - thanks [@djerfy](https://github.com/djerfy)!

## [0.7.1] - 2022-07-02

Howdy farmers! It's been a long while. This release brings everything from `dev` to `main` branch
and going forward, all new development will happen directly on the `main` branch.

There's now an [official docker image](https://github.com/martomi/chiadog/pkgs/container/chiadog/versions) that will
be published automatically with every new release. For the most stable experience, I recommend using the docker image
or sticking to release tagged versions of the `main` branch.

### Added
- Automated docker image builds - thanks [@Artanicus](https://github.com/Artanicus)!
- Pushcut notifier - thanks [@Connor-Knabe](https://github.com/Connor-Knabe)!
- IFTTT notifier - thanks [@Connor-Knabe](https://github.com/Connor-Knabe)!
- Optional notification for plot increases - thanks [@Connor-Knabe](https://github.com/Connor-Knabe)!
- Ability to run in sandboxed systemd service - thanks [@alexking](https://github.com/alexking)!
- Option to disable SMPT Auth - thanks [@unsocial-bleach](https://github.com/unsocial-bleach)!
- Config for log handlers and option to filter dust storm wallet events - thanks [@egeedot](https://github.com/egeedot) and [@martomi](https://github.com/martomi)!

### Fixed
- Thanks [@kanasite](https://github.com/kanasite) for fixing the block parser and adding tests!


## [0.7.0] - 2021-07-29

This release adds support for pooling!

### Added
- Correct handling for partial proof submissions - thanks [@kanasite](https://github.com/kanasite)!
  - No more spam notifications for every partial proof submissions when pooling
  - Notifications about proofs are replaced by found block notifications
  - Daily Stats will report the number of partials and blocks found.
- (Optional) Integration: Grafana annotations - thanks [@pieterhelsen](https://github.com/pieterhelsen)!

### Fixed
- Thanks [@pieterhelsen](https://github.com/pieterhelsen) for restoring compatibility with Python 3.7.
- Thanks [@pieterhelsen](https://github.com/pieterhelsen) for keeping our CI tasks running and fixing type checks dependencies.

## [0.6.1] - 2021-07-14

The purpose of this minor release is to get out a few fixes. No major new features in this release.

### Added
- Config: Now you can specify minutes (hh:mm) for daily stats scheduling (thanks [@kanasite](https://github.com/kanasite))

### Fixed
- Fix corrupt log offset file after unexpected shutdown of chiadog (thanks [@pieterhelsen](https://github.com/pieterhelsen))
- Windows: Fix parsing for finished signage points (thanks [@pieterhelsen](https://github.com/pieterhelsen))
- Windows: Fix check for rotated log files (thanks [@mezzovide](https://github.com/mezzovide))
- Install script: Fix detection for already installed packages (thanks [@bradclow](https://github.com/bradclow))
- Check version: Fix cleanup of git process after execution (thanks [@pieterhelsen](https://github.com/pieterhelsen))

## [0.6.0] - 2021-05-25

### Added

- Found proof notifications (again)
- Configurable frequency for daily stats summary (24h default)
- New flag `--version` to display the current version (thanks [@pieterhelsen](https://github.com/pieterhelsen))
- Scripts to automate launching for Windows (thanks [fiveangle](https://github.com/fiveangle))
- Windows WSL installation instructions (thanks [@ougni](https://github.com/ougni))

### Changed

- Increased the warning threshold for last farming events from 60 to 90 seconds.
- Individual skipped signage points do not trigger notifications anymore. They will be part of daily summary.
- Improved Linux install script by installing common dependencies
  (thanks [@pieterhelsen](https://github.com/pieterhelsen))

### Fixed

- Improved log rotation handling on Windows (thanks [@pieterhelsen](https://github.com/pieterhelsen))
- Misleading notifications for decreasing plots during copying. The warning threshold is now 2.
- Wallet notification formatting - received coins are now displayed without scientific notation.
- More correct handling of skipped signage points - most cases are correctly detected but not all.
  (thanks [@ctrlaltdel](https://github.com/ctrlaltdel) for test-case submission)
- More accurate daily stats about skipped signage points - now covered by unit test.

## [0.5.1] - 2021-05-17

The purpose of this minor release is to get out a few fixes. No new features in this release.

### Fixed

- Another fix for handling of signage points around network forks. Should further reduce false-alarm notifications.
- Correctly parse number of eligible plots when larger than 9 (thanks [@ctrlaltdel](https://github.com/ctrlaltdel))
- Correctly parse added coin events and send out reward notifications (thanks [@skweee](https://github.com/skweee))
- Check potential out-of-bound config values for MQTT integration
  (thanks [@pieterhelsen](https://github.com/pieterhelsen))

## [0.5.0] - 2021-05-13

### Added

- Daily Stats: Now include total searches and the percentage of slow seeks
  (thanks [@pieterhelsen](https://github.com/pieterhelsen))
- (Optional) Notification for received rewards / coins. You need to migrate your `config.yaml` by
  adding `wallet_events: true` to your notifiers. See the `config-example.yaml` for reference.

src/notifier/__init__.py
"""Notifier package responsible for user notification

"""
import json
import logging
import re
import time
import traceback

# std
from abc import ABC, abstractmethod
from dataclasses import dataclass
from json_logic import jsonLogic
from typing import List
from enum import Enum

# Ignore Chiadog alerts about being offline due to entire container just launching in the first 30 minutes
MINIMUM_LAUNCH_SECONDS_BEFORE_ALERTING_ABOUT_BEING_OFFLINE = 30 * 60

class EventPriority(Enum):
    """Event priority dictates how urgently
    the user needs to be notified about it
    """

    LOW = -1
    NORMAL = 0
    HIGH = 1


class EventType(Enum):
    """Events can either be user events
    that are propagated directly to the
    user, or keep-alive events that are
    processed to ensure the system runs
    """

    KEEPALIVE = 0
    USER = 1
    DAILY_STATS = 2
    PLOTDECREASE = 3
    PLOTINCREASE = 4


class EventService(Enum):
    """Even service helps to distinguish
    between similar events for different services
    """

    HARVESTER = 0
    FARMER = 1
    FULL_NODE = 2
    DAILY = 3
    WALLET = 4


@dataclass
class Event:
    type: EventType
    priority: EventPriority
    service: EventService
    message: str


class Notifier(ABC):
    """This abstract class provides common interface for
    any notifier implementation. It should be easy to add
    extensions that integrate with variety of services such as
    Pushover, E-mail, Slack, WhatsApp, etc
    """

    def __init__(self, title_prefix: str, config: dict):
        self._program_launch_time = time.time()
        self._title_prefix = title_prefix
        self._config = config
        self._conn_timeout_seconds = 10
        self._notification_types = [EventType.USER]
        self._notification_services = [EventService.HARVESTER, EventService.FARMER, EventService.FULL_NODE]

        daily_stats = config.get("daily_stats", False)
        wallet_events = config.get("wallet_events", False)
        decreasing_plot_events = config.get("decreasing_plot_events", False)
        increasing_plot_events = config.get("increasing_plot_events", False)
        if daily_stats:
            self._notification_types.append(EventType.DAILY_STATS)
            self._notification_services.append(EventService.DAILY)
        if wallet_events:
            self._notification_services.append(EventService.WALLET)
        if decreasing_plot_events:
            self._notification_types.append(EventType.PLOTDECREASE)
        if increasing_plot_events:
            self._notification_types.append(EventType.PLOTINCREASE)

    def get_title_for_event(self, event):
        icon = ""
        if event.priority == EventPriority.HIGH:
            icon = "🚨"
        elif event.priority == EventPriority.NORMAL:
            icon = "⚠️"
        elif event.priority == EventPriority.LOW:
            icon = "ℹ️"

        return f"{icon} {self._title_prefix} {event.service.name}"

    def should_ignore_event(self, event):
        if event.message and event.message.startswith('Test alert'):
            return False  # Never ignore the 'Test alert' message trigged from Machinaris WebUI when testing configs
        # Automatically ignore Chiadog's spurious "Your harvester appears to be offline!" alerts immediately after a relaunch of container
        # Obviously if the Machinaris container (and thus all farming/harvesting) was just started, there will be a gap in the log... 
        if (self._program_launch_time + MINIMUM_LAUNCH_SECONDS_BEFORE_ALERTING_ABOUT_BEING_OFFLINE) >= time.time():
            if (event.service.name == 'HARVESTER' and event.message.startswith("Your harvester appears to be offline!")) or \
                event.message.startswith("Experiencing networking issues?") or \
                event.message.startswith("Cha-ching!"):
                return True
        # Next only ignore if user has set an "ignore" clause in config.xml for a particular Notifier
        if not "ignore" in self._config:
            return False 
        ignore = self._config["ignore"]
        try:
            # First check for one of type, priority, service, and message as a simple filter
            if 'type' in ignore and ignore['type'] == event.type.name:
                return True
            if 'priority' in ignore and ignore['priority'] == event.priority.name:
                return True
            if 'service' in ignore and ignore['service'] == event.service.name:
                return True
            if 'message' in ignore and re.search(ignore['message'], event.message, re.M|re.I):
                return True
            # Then look for compound ignore clause to invoke json logic
            if 'compound' in ignore:
                rule = json.loads(ignore['compound'])
                data = {   
                    "type" : event.type.name.lower(), 
                    "priority" : event.priority.name.lower(),
                    "service" : event.service.name.lower(), 
                    "message" : event.message
                }
                logging.debug("Rule: {0}".format(json.loads(ignore['compound'])))
                logging.debug("Data: {0}".format(data))
                result = jsonLogic(rule, data)                        
                logging.debug("Result: {0}".format(result))
                return result
        except Exception as ex:
            logging.error("Ignore config '{0}' error {1}".format(ignore, str(ex)))
            traceback.print_exc()
        return False

    def should_allow_event(self, event):
        # Next only allow if user has set an "allow" clause in config.xml for a particular Notifier
        if not "allow" in self._config:
            return True  # By default allow all notifications if no "allow" filter is set
        allow = self._config["allow"]
        try:
            # First check for one of type, priority, service, and message as a simple filter
            if 'type' in allow and allow['type'] == event.type.name:
                return True
            if 'priority' in allow and allow['priority'] == event.priority.name:
                return True
            if 'service' in allow and allow['service'] == event.service.name:
                return True
            if 'message' in allow and re.search(allow['message'], event.message, re.M|re.I):
                return True
            # Then look for compound allow clause to invoke json logic
            if 'compound' in allow:
                rule = json.loads(allow['compound'])
                data = {   
                    "type" : event.type.name.lower(), 
                    "priority" : event.priority.name.lower(),
                    "service" : event.service.name.lower(), 
                    "message" : event.message
                }
                logging.debug("Rule: {0}".format(json.loads(allow['compound'])))
                logging.debug("Data: {0}".format(data))
                result = jsonLogic(rule, data)                        
                logging.debug("Result: {0}".format(result))
                return result
        except Exception as ex:
            logging.error("Allow config '{0}' error {1}".format(allow, str(ex)))
            traceback.print_exc()
        return False

    @abstractmethod
    def send_events_to_user(self, events: List[Event]) -> bool:
        """Implementation specific to the integration"""
        pass
